lib/DependencyProvider.js

```diff
@@ -13,50 +13,190 @@ const syncGetWorker /*: {| get: (string) => string, shutDown: () => void |} */ =
+      // Save the updated onlineVersionsCache to disk.
+      const onlineVersions = Object.fromEntries(this.map.entries());
+      fs.writeFileSync(cachePath, JSON.stringify(onlineVersions));
+    } else {
+      // There was a problem and a package got deleted from the server.
+      this.map = onlineVersionsFromScratch(cachePath, remotePackagesUrl);
+    }
+  }
+
+  getVersions(pkg /*: string */) /*: Array<string> */ {
+    const versions = this.map.get(pkg);
+    return versions === undefined ? [] : versions;
   }
 }
 
-// Solve dependencies with http requests when required.
-function solveOnline(
-  elmJson /*: string */,
-  useTest /*: boolean */,
-  extra /*: { [string]: string } */
-) /*: string */ {
-  updateOnlineVersionsCache();
-  listVersionsMemoCache.clear();
+class OnlineAvailableVersionLister {
+  // Memoization cache to avoid doing the same work twice in list.
+  memoCache /*: Map<string, Array<string>> */ = new Map();
+  onlineCache /*: OnlineVersionsCache */;
+
+  constructor(onlineCache /*: OnlineVersionsCache */) {
+    onlineCache.update();
+    this.onlineCache = onlineCache;
+  }
+
+  list(pkg /*: string */) /*: Array<string> */ {
+    const memoVersions = this.memoCache.get(pkg);
+    if (memoVersions !== undefined) {
+      return memoVersions;
+    }
+    const offlineVersions = readVersionsInElmHomeAndSort(pkg);
+    const allVersionsSet = new Set(this.onlineCache.getVersions(pkg));
+    // Combine local and online versions.
+    for (const version of offlineVersions) {
+      allVersionsSet.add(version);
+    }
+    const allVersions = [...allVersionsSet].sort(semverCompare).reverse();
+    this.memoCache.set(pkg, allVersions);
+    return allVersions;
+  }
+}
+
+class OfflineAvailableVersionLister {
+  // Memoization cache to avoid doing the same work twice in list.
+  cache /*: Map<string, Array<string>> */ = new Map();
+
+  list(pkg /*: string */) /*: Array<string> */ {
+    const memoVersions = this.cache.get(pkg);
+    if (memoVersions !== undefined) {
+      return memoVersions;
+    }
+
+    const offlineVersions = readVersionsInElmHomeAndSort(pkg);
+
+    this.cache.set(pkg, offlineVersions);
+    return offlineVersions;
+  }
+}
+
+function readVersionsInElmHomeAndSort(pkg /*: string */) /*: Array<string> */ {
+  const pkgPath = homePkgPath(pkg);
+  let offlineVersions;
   try {
-    return wasm.solve_deps(
-      elmJson,
-      useTest,
-      extra,
-      fetchElmJsonOnline,
-      listAvailableVersionsOnline
-    );
-  } catch (errorMessage) {
-    throw new Error(errorMessage);
+    offlineVersions = fs.readdirSync(pkgPath);
+  } catch (_) {
+    // The directory doesn't exist or we don't have permissions.
+    // It's fine to catch all cases and return an empty list.
+    offlineVersions = [];
+  }
+
+  offlineVersions.sort(semverCompare);
+
+  // Reverse order of subdirectories to have newest versions first.
+  offlineVersions.reverse();
+  return offlineVersions;
+}
+
+class DependancyProvider {
```

Comment:
This is spelled Depend<strong>e</strong>ncy, isn’t it? (English is not my native language 😊 )